File: server/src/main/java/com/fluffy/exam/application/response/ExamWithAnswersResponse.java
package com.fluffy.exam.application.response;

import java.time.LocalDateTime;
import java.util.List;

public record ExamWithAnswersResponse(
        Long id,
        String title,
        String description,
        String status,
        List<QuestionWithAnswersResponse> questions,
        LocalDateTime createdAt,
        LocalDateTime updatedAt
) {

    public interface QuestionWithAnswersResponse {
        Long id();

        String text();

        String passage();

        String type();
    }

    public record AnswerQuestionWithAnswersResponse(
            Long id,
            String text,
            String passage,
            String type,
            String correctAnswer
    ) implements QuestionWithAnswersResponse {
    }

    public record ChoiceQuestionWithAnswersResponse(
            Long id,
            String text,
            String passage,
            String type,
            List<QuestionOptionWithAnswersResponse> options
    ) implements QuestionWithAnswersResponse {

        public record QuestionOptionWithAnswersResponse(
                Long id,
                String text,
                Boolean isCorrect
        ) {
        }
    }
}

File: server/src/main/kotlin/com/fluffy/exam/application/response/ExamWithAnswersResponse.kt
package com.fluffy.exam.application.response

import java.time.LocalDateTime

data class ExamWithAnswersResponse(
    val id: Long,
    val title: String,
    val description: String,
    val status: String,
    val questions: List<QuestionWithAnswersResponse>,
    val createdAt: LocalDateTime,
    val updatedAt: LocalDateTime,
)

interface QuestionWithAnswersResponse {

    val id: Long
    val text: String
    val passage: String
    val type: String
}

data class AnswerQuestionWithAnswersResponse(
    override val id: Long,
    override val text: String,
    override val passage: String,
    override val type: String,
    val correctAnswer: String?
) : QuestionWithAnswersResponse

data class ChoiceQuestionWithAnswersResponse(
    override val id: Long,
    override val text: String,
    override val passage: String,
    override val type: String,
    val options: List<QuestionOptionWithAnswersResponse>
) : QuestionWithAnswersResponse

data class QuestionOptionWithAnswersResponse(
    val id: Long,
    val text: String,
    val isCorrect: Boolean
)
